cwatch.php

```diff
@@ -252,57 +182,119 @@ public function getEmailTags()
      * @see Module::getModule()
      * @see Module::getModuleRow()
      */
-    public function addService($package, array $vars = null, $parent_package = null, $parent_service = null, $status = 'pending') 
-    {
-        $row = $this->getModuleRow();
-        if (isset($row->meta->username)) {
-            $user = $row->meta->username;
-        }
-        if (isset($row->meta->password)) {
-            $pass = $row->meta->password;
-        }
-        if (isset($package->meta->cwatch_license_type)) {
-            $product = $package->meta->cwatch_license_type;
-        }
-        if (isset($package->meta->cwatch_license_term)){
-            $term = $package->meta->cwatch_license_term;
-        }
-        if (isset($package->meta->cwatch_sandbox)){
-            $sandbox = $package->meta->cwatch_sandbox;
+    public function addService(
+        $package,
+        array $vars = null,
+        $parent_package = null,
+        $parent_service = null,
+        $status = 'pending'
+    ) {
+        Loader::loadModels($this, ['Clients', 'Services']);
+
+        $this->validateService($package, $vars);
+
+        if ($this->Input->errors()) {
+            return;
         }
+
+        // Get cWatch API
+        $api = $this->getApi();
+        $errors = [];
         if (isset($vars['use_module']) && $vars['use_module'] == 'true') {
-            Loader::loadModels($this, ['Clients']);
-            $client = $this->Clients->get($vars['client_id'], false);
-            $firstname = $client->first_name;
-            $lastname = $client->last_name;
-            $email = $client->email;
-            $country = $client->country;
-            $this->loadApi($user, $pass, $sandbox);
-            try{
-                $response = $this->api->createlicence($email, $firstname, $lastname, $country, $product, $term);
-                $json = json_decode($response->resp);
+            $response = null;
+            try {
+                // Log request data
+                $this->log('createcommand', serialize($vars), 'input', true);
```

Comment:
Ideally, we'd like to log only information that we're sending in the API request rather than all data submitted in $vars, especially if there is sensitive info (e.g. password) that should not be logged.

Reply:
What would probably help out the most is for the CwatchApi class to have a method to retrieve the last request

```
private $last_request = [];

private function apiRequest(...) 
{
    $this->lastRequest = ['content' => ..., 'headers' => ''];
}
public function lastRequest()
{
    return $this->last_request;
}
```